python-package/xgboost/spark/data.py

```diff
@@ -1,194 +1,207 @@
+        label = concat_or_none(values.get(alias.label, None))
+        weight = concat_or_none(values.get(alias.weight, None))
+        margin = concat_or_none(values.get(alias.margin, None))
+        return DMatrix(
+            data=data, label=label, weight=weight, base_margin=margin, **kwargs
+        )
+
+    train = make(train_data)
+    valid = make(valid_data) if len(valid_data) != 0 else None
+
+    assert train.num_col() == n_features
+    if valid is not None:
+        assert valid.num_col() == train.num_col()
+
+    return train, valid
+
+
+class PartIter(DataIter):
+    """Iterator for creating Quantile DMatrix from partitions."""
+
+    def __init__(self, data: Dict[str, List], on_device: bool) -> None:
+        self._iter = 0
+        self._cuda = on_device
+        self._data = data
+
+        super().__init__()
+
+    def _fetch(self, data: Optional[Sequence[pd.DataFrame]]) -> Optional[pd.DataFrame]:
+        if not data:
+            return None
+
+        if self._cuda:
+            import cudf  # pylint: disable=import-error
+
+            return cudf.DataFrame(data[self._iter])
+
+        return data[self._iter]
+
+    def next(self, input_data: Callable) -> int:
+        if self._iter == len(self._data[alias.data]):
+            return 0
+        input_data(
+            data=self._fetch(self._data[alias.data]),
+            label=self._fetch(self._data.get(alias.label, None)),
+            weight=self._fetch(self._data.get(alias.weight, None)),
+            base_margin=self._fetch(self._data.get(alias.margin, None)),
         )
-        val_dmatrix = DMatrix(
-            data=val_x,
-            label=val_y,
-            weight=val_w,
-            base_margin=val_b_m,
-            **dmatrix_kwargs,
+        self._iter += 1
+        return 1
+
+    def reset(self) -> None:
+        self._iter = 0
+
+
+def create_qdm_from_partitions(
+    iterator: Iterator[pd.DataFrame],
+    features_cols: Sequence[str],
+    **kwargs: Any,
+) -> Tuple[DMatrix, Optional[DMatrix]]:
+    """Create a Quantile DMatrix from pyspark partitions."""
+    if not features_cols:
+        raise ValueError("features cols must not be empty.")
+
+    train_data: Dict[str, List] = defaultdict(list)
+    valid_data: Dict[str, List] = defaultdict(list)
+    n_features: int = 0
+
+    def append(part: pd.DataFrame, name: str, is_valid: bool) -> None:
+        nonlocal n_features
+        if name == alias.data or name in part.columns:
+            if name == alias.data:
+                cname = features_cols
+            else:
+                cname = name
+
+            array = part[cname]
+            if name == alias.data:
+                if n_features == 0:
+                    n_features = array.shape[1]
+                assert n_features == array.shape[1]
+
+            if is_valid:
+                valid_data[name].append(array)
+            else:
+                train_data[name].append(array)
+
+    cache_partitions(iterator, append)
```

Comment:
For DeviceQuantileDMatrix, it supports iterator as input, we don't need to load the whole partition data into memory, right ?

Reply:
Current code (for GPU) is:
1. load whole partition data into memory arrays
2. construct an iterator from memory arrays in step (1)
3. construct DMatrix from iterator in step (2)

What I propose it not to write data to disk, but change step (1)(2) into:  construct an iterator from spark python UDF data iterator. This avoid load data into memory in (1) step.

We can do this improvement in follow-up PR.

Reply:
@WeichenXu123 I think the data is still loaded into memory. We need to iterate through the data twice (notice the reset function). Unless the spark iterator can be reset and start again, we need to cache the data somewhere.

Reply:
ah, got it , spark udf data iterator does not support reset. that's bad.

Reply:
Previously I thought DMatrix should accept an one-shot iterator, it exhausts the iterator and cache all data into DMatrix format in memory, and then xgboost training can use the DMatrix data in mutilple passes. Why DMatrix iteration API is not designed this way ?

Reply:
@WeichenXu123 Apologies for the confusion here. Here's a little bit of clarification on the difference between different DMatric constructors:
- Normal DMatrix: `DMatrix(np.array(...))`
  It accepts one single batch of data, and constructs an internal CSR representation that can be used by all algorithms. It has 2 issues with distributed training: firstly, it needs to concatenate all partitions, which doubles the memory usage, secondly, it's a csr, which can triple the memory usage when input is dense.
- `QuantileDMatrix`: This is designed to be as "inplace" as possible and can only be used by GPU hist.  It omits the CSR representation and constructs the histogram index directly from external data. The iterator concept you see there is designed for distributed system that handles data in terms of partitions, the sole purpose of accepting an iterator instead of a whole blob here is to avoid the `concatentate` function on the input data, which doubles the memory. (`concat_or_none` in this PR.)
- External memory: `DMatrix(iterator)`. Yes, using CPU `hist`/`approx` DMatrix caches the batches on disk and iterates through it. But it's quite slow since we need to iterate through the data 3-4 times for each layer of the tree, you can estimate the time usage based on the throughput of your hard disk.

Reply:
Got it.

so QuantileDMatrix this does not support sparse data input ?

Reply:
I would like to assign all data iterator related work to you because you have deeper understanding on this. :)

Reply:
> so QuantileDMatrix this does not support sparse data input ?

It will suppotr CSR once the CPU impl is merged.

> I would like to assign all data iterator related work to you because you have deeper understanding on this. :)

Got it.